controller/internal/controller/consumer_controller.go

```diff
@@ -181,3 +201,76 @@ func (r *ConsumerReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		)
 	return controller.Complete(r)
 }
+
+// checkConsumerConflicts Perform full conflict detection and rebuild the index
+func (r *ConsumerReconciler) checkConsumerConflicts(ctx context.Context, state *consumerReconcileState) error {
+	r.keyIndex.mu.Lock()
+	defer r.keyIndex.mu.Unlock()
+
+	// Clear old indexes
+	r.keyIndex.index = make(map[string]map[string]map[string]string)
+
+	// Check conflicts for all valid consumers in the current state
+	for ns, consumers := range state.namespaceToConsumers {
+		for _, consumer := range consumers {
+			if err := r.indexConsumer(ns, consumer); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// indexConsumer method to use the plugin's Index method
+func (r *ConsumerReconciler) indexConsumer(namespace string, consumer *mosniov1.Consumer) error {
+	for pluginName, plugin := range consumer.Spec.Auth {
+		// Get the plugin from the registry
+		registeredPlugin, exists := pluginRegistry[pluginName]
+		if !exists {
+			return fmt.Errorf("unknown plugin: %s", pluginName)
+		}
+
+		// Create configuration instances through the plugin interface
+		pluginConfig := registeredPlugin.NewConfig()
+		if err := json.Unmarshal(plugin.Config.Raw, pluginConfig); err != nil {
+			return fmt.Errorf("invalid config for plugin %s: %w", pluginName, err)
+		}
+
+		key := pluginConfig.Index()
+
+		// Initialize the three-level map if needed
+		if r.keyIndex.index[namespace] == nil {
+			r.keyIndex.index[namespace] = make(map[string]map[string]string)
+		}
+		if r.keyIndex.index[namespace][pluginName] == nil {
+			r.keyIndex.index[namespace][pluginName] = make(map[string]string)
+		}
+
+		// Check for collision
+		if existing, exists := r.keyIndex.index[namespace][pluginName][key]; exists {
+			return fmt.Errorf(
+				"key conflict in namespace %s: plugin=%s key=%s (already used by %s)",
+				namespace, pluginName, key, existing,
+			)
+		}
+		r.keyIndex.index[namespace][pluginName][key] = consumer.Name
+	}
+	return nil
+}
+
+// Plugin Define the plugin interface
+type Plugin interface {
```

Comment:
Please reuse the interface in https://github.com/mosn/htnn/blob/main/api/pkg/plugins/type.go